File: Assets/src/model/enemies/Thistlebog.cs
using System.Collections.Generic;
using UnityEngine;

[System.Serializable]
[ObjectInfo(description: "Moves slowly.\nSummons a ring of Brambles around you that disappear after 10 turns (needs vision).\nInterrupted when taking damage.", flavorText: "Four limbs and a face are the only bits of humanity left in this husk.")]
public class Thistlebog : AIActor, ITakeAnyDamageHandler {
  public static new ActionCosts StaticActionCosts = new ActionCosts(Actor.StaticActionCosts) {
    [ActionType.MOVE] = 2f,
  };
  protected override ActionCosts actionCosts => StaticActionCosts;
  public override float turnPriority => 50;

  private float cooldown = 0;
  public Thistlebog(Vector2Int pos) : base(pos) {
    faction = Faction.Enemy;
    hp = baseMaxHp = 10;
  }

  internal override (int, int) BaseAttackDamage() => (0, 0);

  public override float Step() {
    var dt = base.Step();
    if (cooldown > 0) {
      cooldown -= dt;
    }
    return dt;
  }

  public void HandleTakeAnyDamage(int damage) {
    if (damage > 0 && task is TelegraphedTask) {
      ClearTasks();
      SetTasks(new WaitTask(this, 1));
      statuses.Add(new SurprisedStatus());
    }
  }

  protected override ActorTask GetNextTask() {
    if (!CanTargetPlayer()) {
      return new MoveRandomlyTask(this, AvoidBrambles);
    }
    var player = GameModel.main.player;
    if (IsNextTo(player)) {
      return new RunAwayTask(this, player.pos, 1, false);
    }
    if (cooldown > 0) {
      return new MoveRandomlyTask(this, AvoidBrambles);
    }
    if (Util.DiamondDistanceToPlayer(this) <= 2) {
      return new TelegraphedTask(this, 1, new GenericBaseAction(this, SummonBramblesAroundPlayer));
    } else {
      return new ChaseTargetTask(this, GameModel.main.player, 1).MaxMoves(1);
    }
  }

  public static bool AvoidBrambles(Tile t) => !(t.grass is Brambles);

  private void SummonBramblesAroundPlayer() {
    if (CanTargetPlayer() && cooldown <= 0) {
      cooldown = 10;
      var center = GameModel.main.player.pos;
      foreach (var tile in floor.GetAdjacentTiles(center)) {
        if (Brambles.CanOccupy(tile) && tile.pos != center) {
          var brambles = new Brambles(tile.pos);
          floor.Put(brambles);
          brambles.AddTimedEvent(10, brambles.KillSelf);
        }
      }
    }
  }
}

File: web/src/model/enemies/Thistlebog.ts
import { AIActor } from './AIActor';
import { ActorTask } from '../ActorTask';
import { ChaseTargetTask } from '../tasks/ChaseTargetTask';
import { MoveRandomlyTask } from '../tasks/MoveRandomlyTask';
import { RunAwayTask } from '../tasks/RunAwayTask';
import { WaitTask } from '../tasks/WaitTask';
import { TelegraphedTask } from '../tasks/TelegraphedTask';
import { GenericBaseAction, ActionCosts } from '../BaseAction';
import { SurprisedStatus } from '../tasks/SleepTask';
import { Vector2Int } from '../../core/Vector2Int';
import { ActionType } from '../../core/types';
import { GameModelRef } from '../GameModelRef';
import { TAKE_ANY_DAMAGE_HANDLER, type ITakeAnyDamageHandler } from '../Body';
import { entityRegistry } from '../../generator/entityRegistry';

/**
 * Moves slowly. Summons Brambles around the player (needs vision).
 * Interrupted when taking damage.
 * Port of C# Thistlebog.cs.
 */
export class Thistlebog extends AIActor implements ITakeAnyDamageHandler {
  readonly [TAKE_ANY_DAMAGE_HANDLER] = true as const;
  private cooldown = 0;

  get turnPriority(): number {
    return 50;
  }

  constructor(pos: Vector2Int) {
    super(pos);
    this._hp = this._baseMaxHp = 10;
  }

  protected get actionCosts(): ActionCosts {
    const costs = ActionCosts.default();
    costs.set(ActionType.MOVE, 2);
    return costs;
  }

  baseAttackDamage(): [number, number] {
    return [0, 0];
  }

  step(): number {
    const dt = super.step();
    if (this.cooldown > 0) {
      this.cooldown -= dt;
    }
    return dt;
  }

  handleTakeAnyDamage(damage: number): void {
    if (damage > 0 && this.task instanceof TelegraphedTask) {
      this.clearTasks();
      this.setTasks(new WaitTask(this, 1));
      this.statuses.add(new SurprisedStatus());
    }
  }

  protected getNextTask(): ActorTask {
    if (!this.canTargetPlayer()) {
      return new MoveRandomlyTask(this);
    }
    const player = GameModelRef.main.player;
    if (this.isNextTo(player)) {
      return new RunAwayTask(this, player.pos, 1, false);
    }
    if (this.cooldown > 0) {
      return new MoveRandomlyTask(this);
    }
    // Diamond distance to player
    const dx = Math.abs(this.pos.x - player.pos.x);
    const dy = Math.abs(this.pos.y - player.pos.y);
    if (dx + dy <= 2) {
      return new TelegraphedTask(this, 1, new GenericBaseAction(this, () => this.summonBramblesAroundPlayer()));
    }
    const chase = new ChaseTargetTask(this, player, 1);
    chase.setMaxMoves(1);
    return chase;
  }

  private summonBramblesAroundPlayer(): void {
    if (!this.canTargetPlayer() || this.cooldown > 0) return;
    this.cooldown = 10;
    // TODO: Summon Brambles grass around player when Brambles is ported
    // For now, this is a stub. When Brambles exists:
    // const center = GameModelRef.main.player.pos;
    // for (const tile of this.floor!.getAdjacentTiles(center)) {
    //   if (Brambles.canOccupy(tile) && !tile.pos.equals(center)) {
    //     const brambles = new Brambles(tile.pos);
    //     this.floor!.put(brambles);
    //     brambles.addTimedEvent(10, brambles.killSelf);
    //   }
    // }
  }
}

entityRegistry.register('Thistlebog', Thistlebog);
